File: routing/instance.py
"""Names in, arrays out.

Everything downstream of this module works in integers: a port is an index
into 0..n_ports, a task is an index into 0..n_tasks, and every cost is an
int32 tick count in an array. This is the only place that knows a port has a
name, that a level gates anything, or that any of it came off a TSV.

Build one with `Instance.at_level(30)`.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from . import params as params_mod
from .errors import ModelError
from .params import Params
from .portmatrix import PortMatrix
from .world import World

NONE = -1  # the empty slot, used for absent ports, tasks and offers alike


@dataclass(frozen=True)
class Instance:
    """The static problem, as arrays. Nothing here changes during an episode.

    Ragged per-board task pools are held CSR-style: the pool of board `p` is
    `pool[pool_ptr[p]:pool_ptr[p + 1]]`.
    """
    level: int
    params: Params

    # ports
    sail: np.ndarray        # (P, P) int32, ticks, diagonal 0
    charter: np.ndarray     # (P,)   int32, ticks, NONE where no charter ship
    recall: np.ndarray      # (P,)   int32, ticks, NONE where no shipwright
    has_board: np.ndarray   # (P,)   bool

    # tasks
    task_board: np.ndarray     # (T,) int32
    task_origin: np.ndarray    # (T,) int32
    task_dest: np.ndarray      # (T,) int32
    task_xp: np.ndarray        # (T,) int32
    task_eligible: np.ndarray  # (T,) bool - may be accepted at this level

    # per-board pools, CSR
    pool_ptr: np.ndarray    # (P + 1,) int32
    pool: np.ndarray        # (sum,)   int32

    # names, for reporting only - the simulator never reads these
    port_names: tuple[str, ...]
    task_names: tuple[str, ...]

    @property
    def n_ports(self) -> int:
        return len(self.port_names)

    @property
    def n_tasks(self) -> int:
        return len(self.task_names)

    @property
    def capacity(self) -> int:
        return self.params.capacity_at(self.level)

    def board_pool(self, port: int) -> np.ndarray:
        return self.pool[self.pool_ptr[port]:self.pool_ptr[port + 1]]

    @classmethod
    def at_level(cls, level: int, world: World | None = None,
                 matrix: PortMatrix | None = None, params: Params | None = None) -> Instance:
        world = world or World.load()
        matrix = matrix or PortMatrix.load()
        params = params or Params.load()
        charter = params_mod.charter_ports()

        names = tuple(sorted(n for n, p in world.ports.items() if p.dock_level <= level))
        index = {name: i for i, name in enumerate(names)}
        size = len(names)
        if not size:
            raise ModelError(f'no port is dockable at level {level}')

        sail = np.zeros((size, size), np.int32)
        for i, a in enumerate(names):
            for j, b in enumerate(names):
                if i != j:
                    sail[i, j] = round(matrix.between(a, b) / params.sail_speed) + params.t_dock

        charter_ticks = np.full(size, NONE, np.int32)
        recall_ticks = np.full(size, NONE, np.int32)
        has_board = np.zeros(size, bool)
        for name, i in index.items():
            port = world.ports[name]
            has_board[i] = port.notice_board
            if port.shipwright:
                recall_ticks[i] = params.t_recall
            if name in charter:
                charter_ticks[i] = params.t_charter

        # A board offers from its whole pool regardless of level, so tasks that
        # cannot be done still occupy an offer slot. Keeping them is the point:
        # dropping them would quietly inflate how much choice a board gives.
        chosen = [t for t in world.tasks.values() if t.board in index]
        chosen.sort(key=lambda t: t.id)
        task_index = {t.id: i for i, t in enumerate(chosen)}

        def port_or_none(name: str) -> int:
            return index.get(name, NONE)

        board = np.array([index[t.board] for t in chosen], np.int32)
        origin = np.array([port_or_none(t.origin) for t in chosen], np.int32)
        dest = np.array([port_or_none(t.destination) for t in chosen], np.int32)
        xp = np.array([t.xp or 0 for t in chosen], np.int32)
        eligible = np.array([t.level <= level and t.origin in index and t.destination in index
                             and t.xp is not None for t in chosen], bool)

        order = np.argsort(board, kind='stable')
        pool = order.astype(np.int32)
        pool_ptr = np.zeros(size + 1, np.int32)
        np.cumsum(np.bincount(board, minlength=size), out=pool_ptr[1:])

        return cls(
            level=level, params=params,
            sail=sail, charter=charter_ticks, recall=recall_ticks, has_board=has_board,
            task_board=board, task_origin=origin, task_dest=dest, task_xp=xp,
            task_eligible=eligible, pool_ptr=pool_ptr, pool=pool,
            port_names=names,
            task_names=tuple(f'{t.cargo} x{t.qty} {t.origin}->{t.destination}' for t in chosen))

    def describe(self) -> str:
        boards = int(self.has_board.sum())
        live = int(self.task_eligible.sum())
        pools = np.diff(self.pool_ptr)[self.has_board]
        return (f'level {self.level}: {self.n_ports} ports, {boards} boards, '
                f'{self.n_tasks} tasks ({live} eligible), capacity {self.capacity}, '
                f'{self.params.courier_per_board} offers from pools of '
                f'{pools.min()}-{pools.max()}')

File: search/src/instance.rs
//! The static problem, as flat arrays: what `porttasks.routing.problem.instance`
//! computes, read from the JSON that `..export` writes.
//!
//! Nothing here parses a table. `tables/` is the one source of truth and Python
//! owns it; this end only reads what Python already derived, so the two cannot
//! drift into disagreeing about what a port costs.
//!
//! Everything is an index. A port is `0..n_ports`, a task is `0..n_tasks`, and
//! `NONE` is the absent one - an empty slot, an unreachable port, a task that
//! does not exist. Ticks are `i32` throughout, as they are in Python: the
//! game's own quantum, so the arithmetic is exact.

use std::fs;
use std::path::{Path, PathBuf};

use serde::Deserialize;

/// The absent port, task or offer. Matches `instance.NONE` on the Python side.
pub const NONE: i32 = -1;

/// The format `export.py` writes. A mismatch means the two halves were built
/// from different revisions, which is worth refusing rather than guessing at.
const FORMAT: u32 = 2;

#[derive(Debug, Deserialize)]
pub struct Params {
    pub courier_per_board: usize,
    pub reroll_completions: u32,
    pub t_board: i32,
    pub t_drop: i32,
}

#[derive(Debug, Deserialize)]
struct Ports {
    names: Vec<String>,
    sail: Vec<Vec<i32>>,
    /// Ticks to reach without the boat - charter ship or magic teleport alike.
    travel: Vec<i32>,
    recall: Vec<i32>,
    has_board: Vec<bool>,
}

#[derive(Debug, Deserialize)]
struct Tasks {
    names: Vec<String>,
    board: Vec<i32>,
    origin: Vec<i32>,
    dest: Vec<i32>,
    xp: Vec<i32>,
    eligible: Vec<bool>,
}

#[derive(Debug, Deserialize)]
struct Pools {
    ptr: Vec<i32>,
    flat: Vec<i32>,
}

#[derive(Debug, Deserialize)]
struct Wire {
    format: u32,
    level: u32,
    capacity: usize,
    params: Params,
    ports: Ports,
    tasks: Tasks,
    pools: Pools,
}

/// The static problem at one level. Immutable for the life of a solve.
///
/// The sail matrix is stored row-major and flat rather than as `Vec<Vec<i32>>`:
/// it is read once per simulated action and wants to be one cache line away,
/// not two pointer hops.
#[derive(Debug)]
pub struct Instance {
    pub level: u32,
    pub capacity: usize,
    pub params: Params,

    pub n_ports: usize,
    sail: Vec<i32>, // n_ports * n_ports, row-major
    pub travel: Vec<i32>,
    pub recall: Vec<i32>,
    pub has_board: Vec<bool>,

    pub n_tasks: usize,
    pub task_board: Vec<i32>,
    pub task_origin: Vec<i32>,
    pub task_dest: Vec<i32>,
    pub task_xp: Vec<i32>,
    pub task_eligible: Vec<bool>,

    pool_ptr: Vec<i32>,
    pool: Vec<i32>,

    pub port_names: Vec<String>,
    pub task_names: Vec<String>,
}

impl Instance {
    /// Sail time in ticks. Includes docking and cargo handling, as Python's does.
    #[inline]
    pub fn sail(&self, from: usize, to: usize) -> i32 {
        self.sail[from * self.n_ports + to]
    }

    /// The tasks board `port` may offer. Empty where the port has no board.
    #[inline]
    pub fn board_pool(&self, port: usize) -> &[i32] {
        let (lo, hi) = (
            self.pool_ptr[port] as usize,
            self.pool_ptr[port + 1] as usize,
        );
        &self.pool[lo..hi]
    }

    pub fn boards(&self) -> impl Iterator<Item = usize> + '_ {
        (0..self.n_ports).filter(|&p| self.has_board[p])
    }

    pub fn load(path: impl AsRef<Path>) -> Result<Instance, String> {
        let path = path.as_ref();
        let text = fs::read_to_string(path).map_err(|e| format!("{}: {e}", path.display()))?;
        let wire: Wire =
            serde_json::from_str(&text).map_err(|e| format!("{}: {e}", path.display()))?;
        if wire.format != FORMAT {
            return Err(format!(
                "{}: format {} but this build reads {FORMAT}; rerun \
                 `python3 -m porttasks.routing.problem.export`",
                path.display(),
                wire.format
            ));
        }
        Ok(Instance::from(wire))
    }

    /// Load the export for `level` from the repo's `derived/`.
    pub fn at_level(level: u32) -> Result<Instance, String> {
        Instance::load(derived().join(format!("instance_l{level}.json")))
    }

    /// The one line `Instance.describe()` prints in Python, to the character.
    /// Cheap, and it is how we check the seam did not lose anything.
    pub fn describe(&self) -> String {
        let boards = self.has_board.iter().filter(|&&b| b).count();
        let live = self.task_eligible.iter().filter(|&&e| e).count();
        let sizes: Vec<i32> = self
            .boards()
            .map(|p| self.pool_ptr[p + 1] - self.pool_ptr[p])
            .collect();
        let (lo, hi) = (
            sizes.iter().min().copied().unwrap_or(0),
            sizes.iter().max().copied().unwrap_or(0),
        );
        format!(
            "level {}: {} ports, {boards} boards, {} tasks ({live} eligible), \
             capacity {}, {} offers from pools of {lo}-{hi}",
            self.level, self.n_ports, self.n_tasks, self.capacity, self.params.courier_per_board
        )
    }
}

impl From<Wire> for Instance {
    fn from(w: Wire) -> Instance {
        let n_ports = w.ports.names.len();
        Instance {
            level: w.level,
            capacity: w.capacity,
            params: w.params,
            n_ports,
            sail: w.ports.sail.into_iter().flatten().collect(),
            travel: w.ports.travel,
            recall: w.ports.recall,
            has_board: w.ports.has_board,
            n_tasks: w.tasks.names.len(),
            task_board: w.tasks.board,
            task_origin: w.tasks.origin,
            task_dest: w.tasks.dest,
            task_xp: w.tasks.xp,
            task_eligible: w.tasks.eligible,
            pool_ptr: w.pools.ptr,
            pool: w.pools.flat,
            port_names: w.ports.names,
            task_names: w.tasks.names,
        }
    }
}

/// The repo's `derived/`, found from the crate rather than the working
/// directory - `search/` sits one level below the root, exactly as
/// `porttasks.paths` anchors the Python half.
pub fn derived() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("the crate is a directory inside the repo")
        .join("derived")
}

#[cfg(test)]
mod tests {
    use super::*;

    fn level_30() -> Instance {
        Instance::at_level(30).expect("run `python3 -m porttasks.routing.problem.export` first")
    }

    #[test]
    fn describes_itself_as_python_does() {
        assert_eq!(
            level_30().describe(),
            "level 30: 10 ports, 9 boards, 171 tasks (98 eligible), capacity 3, \
             5 offers from pools of 19-19"
        );
    }

    #[test]
    fn sail_matrix_is_symmetric_with_zero_diagonal() {
        let inst = level_30();
        for a in 0..inst.n_ports {
            assert_eq!(inst.sail(a, a), 0);
            for b in 0..inst.n_ports {
                assert_eq!(inst.sail(a, b), inst.sail(b, a));
            }
        }
    }

    #[test]
    fn every_board_can_fill_its_offers() {
        // the draw is without replacement, so a short pool would be unsatisfiable
        let inst = level_30();
        for port in inst.boards() {
            assert!(inst.board_pool(port).len() >= inst.params.courier_per_board);
        }
    }

    #[test]
    fn pools_keep_tasks_the_level_cannot_do() {
        // a board draws from its whole pool, so an ineligible task still burns
        // an offer slot; dropping them would inflate how much choice a board gives
        let inst = level_30();
        assert!(inst
            .boards()
            .flat_map(|p| inst.board_pool(p))
            .any(|&t| !inst.task_eligible[t as usize]));
    }
}
